internal/datanode/channel_checkpoint_updater.go

```diff
@@ -122,24 +149,43 @@ func (ccu *channelCheckpointUpdater) execute() {
 	})
 }
 
-func (ccu *channelCheckpointUpdater) addTask(channelPos *msgpb.MsgPosition, callback func()) {
+func (ccu *channelCheckpointUpdater) execute() {
+	ccu.mu.RLock()
+	tasks := lo.Values(ccu.tasks)
+	ccu.mu.RUnlock()
+
+	ccu.updateCheckpoints(tasks)
+}
+
+func (ccu *channelCheckpointUpdater) AddTask(channelPos *msgpb.MsgPosition, flush bool, callback func()) {
 	if channelPos == nil || channelPos.GetMsgID() == nil || channelPos.GetChannelName() == "" {
 		log.Warn("illegal checkpoint", zap.Any("pos", channelPos))
 		return
 	}
+	if flush {
+		defer ccu.trigger()
+	}
 	channel := channelPos.GetChannelName()
-	ccu.mu.RLock()
-	if ccu.tasks[channel] != nil && channelPos.GetTimestamp() <= ccu.tasks[channel].pos.GetTimestamp() {
-		ccu.mu.RUnlock()
+	task, ok := ccu.getTask(channelPos.GetChannelName())
+	if !ok {
+		ccu.mu.Lock()
+		defer ccu.mu.Unlock()
+		ccu.tasks[channel] = &channelCPUpdateTask{
+			pos:      channelPos,
+			callback: callback,
+			flush:    flush,
+		}
 		return
 	}
-	ccu.mu.RUnlock()
 
-	ccu.mu.Lock()
-	defer ccu.mu.Unlock()
-	ccu.tasks[channel] = &channelCPUpdateTask{
-		pos:      channelPos,
-		callback: callback,
+	if task.pos.GetTimestamp() < channelPos.GetTimestamp() || (flush && !task.flush) {
```

Comment:
This may cause the channel checkpoint to regress. Perhaps this approach might be better:
```
if task.pos.GetTimestamp() < channelPos.GetTimestamp() {
  ccu.mu.Lock()
  defer ccu.mu.Unlock()
  ccu.tasks[channel] = &channelCPUpdateTask{
	  pos:      channelPos,
	  callback: callback,
	  flush:    flush || task.flush,
  }
  return
}
if flush && !task.flush {
  ccu.mu.Lock()
  defer ccu.mu.Unlock()
  ccu.tasks[channel].flush = true
}
```

Reply:
fixed